gcsio/src/main/java/com/google/cloud/hadoop/gcsio/GoogleCloudStorageClientImpl.java

```diff
@@ -550,6 +566,201 @@ public void deleteBuckets(List<String> bucketNames) throws IOException {
     }
   }
 
+  /** See {@link GoogleCloudStorage#getItemInfos(List)} for details about expected behavior. */
+  @Override
+  public List<GoogleCloudStorageItemInfo> getItemInfos(List<StorageResourceId> resourceIds)
+      throws IOException {
+    logger.atFiner().log("getItemInfos(%s)", resourceIds);
+
+    if (resourceIds.isEmpty()) {
+      return new ArrayList<>();
+    }
+
+    Map<StorageResourceId, GoogleCloudStorageItemInfo> itemInfos = new ConcurrentHashMap<>();
+    Set<IOException> innerExceptions = newConcurrentHashSet();
+    BatchExecutor executor = new BatchExecutor(storageOptions.getBatchThreads());
+    // For each resourceId, we'll either directly add ROOT_INFO, enqueue a Bucket fetch request,
+    // or enqueue a StorageObject fetch request.
+    try {
+      for (StorageResourceId resourceId : resourceIds) {
+        if (resourceId.isRoot()) {
+          itemInfos.put(resourceId, GoogleCloudStorageItemInfo.ROOT_INFO);
+        } else if (resourceId.isBucket()) {
+          executor.queue(
+              () -> getBucket(resourceId.getBucketName()),
+              getBucketCallback(resourceId, innerExceptions, itemInfos));
+
+        } else if (resourceId.isStorageObject()) {
+          executor.queue(
+              () -> getBlob(resourceId), getBlobCallback(resourceId, innerExceptions, itemInfos));
+        }
+      }
+    } finally {
+      executor.shutdown();
+    }
+    if (!innerExceptions.isEmpty()) {
+      throw GoogleCloudStorageExceptions.createCompositeException(innerExceptions);
+    }
+
+    // Assemble the return list in the same order as the input arguments.
+    List<GoogleCloudStorageItemInfo> sortedItemInfos = new ArrayList<>();
+    for (StorageResourceId resourceId : resourceIds) {
+      checkState(
+          itemInfos.containsKey(resourceId),
+          "Somehow missing resourceId '%s' from map: %s",
+          resourceId,
+          itemInfos);
+      sortedItemInfos.add(itemInfos.get(resourceId));
```

Comment:
What will happen if getting this resourceId failed? it will not be present in the map, correct?

Reply:
In cases where fetching the resource failed, the exceptions will be added as part of `innerExceptions` https://github.com/GoogleCloudDataproc/hadoop-connectors/pull/1089/files/b118ae14967f9a6299f10c87e6b8ee24b17787ec#diff-23a518b24b3819c111aacec5ec168906787ef291585a52e128c1ce776b8ed0fbR601 and these lines wont be executed.